### PR TITLE
Adds ZManaged missing provide methods

### PR DESCRIPTION
**Pull request by saraiva132**
Using ManagedApp and kinda feel it would be a nice addition. Updated the documentation to be similar to ZIO methods. 

Adds `provideLayer`, `provideManaged`, `provideManagedSome` and `provideSomeM` to `ZManaged`

### CODE REVIEW

**Comment by adamgfraser** [comment]
I think this is a great addition. While we're at it do we want to add the rest of the usual cast of characters of `provide` variants? With this PR we would have `provideSomeM` but not `provideM` so I would think we should at least provide that. There is also `provideManaged` and `provideSomeManaged`. Not just if there would be any special issues with providing a managing resource to another managed. There is also the new `provideLayer`.

**Comment by saraiva132** [comment]
Since reservation release does not accept an effect that can throw errors. I had to change things a bit. Before adding more combinators, what do you guys think. @adamgfraser @mschuwalow

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 1]
```diff
@@ -618,10 +618,42 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
...
+   * {{{
+   * val effect: ZManaged[Console with Logging, Nothing, Unit] = ???
+   *
+   * val r0: URIO[Console, Console with Logging] = ???
+   *
+   * effect.provideSomeM(r0)
+   * }}}
+   */
+  def provideSomeM[R0, E1 >: E](
+    f: ZIO[R0, E1, R]
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZManaged.scala]: I think we should say that the effect cannot fail here.

Comment (saraiva132) [reply core/shared/src/main/scala/zio/ZManaged.scala]: I thought about that. I thought about going that route. I was wondering if it was too restrictive. But then again, accepting an effect that can fail, with a die hidden under the hood for the release, might lead to even yet more unexpected results, plus, it is binding to a subtype of throwable.

Comment (adamgfraser) [reply core/shared/src/main/scala/zio/ZManaged.scala]: Actually there is some weirdness here because we are running the effect twice. Why don't we implement this as `f.toManaged_.flatMap(self.provide)`? Then we can also remove the constraint on the error type.

**Comment by saraiva132** [comment]
Today I was learning about `ZLayer` and `Has` and I was wondering how this translate mechanism actually works. `provideLayer` compiles without it which kinda scares me. I might be missing something but by abstracting this behaviour shared by both `provideLayer` into this translate method I could not make it work with by passing the type explicitly  or making the type constraints I wanted . Most likely I am not making any sense in the first place but I would like a second opinion. Thanks

File: core/shared/src/main/scala/zio/ZManaged.scala (hunk 1 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/ZManaged.scala]: Looks great now! Very minor suggestion, maybe call the parameter `r0` instead of `f` for consistency with same method on ZIO?

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 2]
```diff
@@ -612,16 +612,80 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
...
+   * Provides a layer to the `ZManaged`, which translates it to another level.
+   */
+  def provideLayer[E1 >: E, R0, R1 <: Has[_]](layer: ZLayer[R0, E1, R1])(implicit ev: R1 <:< R): ZManaged[R0, E1, A] =
+    layer.translate.value.flatMap(self.provide(_))
+
+  /**
+   * An effectual version of `provide`, useful when the act of provision
+   * requires an effect.
+   */
+  def provideM[E1 >: E](f: ZIO[Any, E1, R])(implicit ev: NeedsEnv[R]): Managed[E1, A] =
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZManaged.scala]: Maybe call parameter `r` instead of `f`?

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 3]
```diff
@@ -612,16 +612,80 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
...
    */
-  def provide(r: R)(implicit ev: NeedsEnv[R]): ZManaged[Any, E, A] =
+  def provide(r: R)(implicit ev: NeedsEnv[R]): Managed[E, A] =
     provideSome(_ => r)
 
+  /**
+   * Provides a layer to the `ZManaged`, which translates it to another level.
+   */
+  def provideLayer[E1 >: E, R0, R1 <: Has[_]](layer: ZLayer[R0, E1, R1])(implicit ev: R1 <:< R): ZManaged[R0, E1, A] =
+    layer.translate.value.flatMap(self.provide(_))
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZManaged.scala]: How about `provideSomeManaged(layer.value.map(ev))`? Then we can revert the changes to `ZIO.scala` and `ZLayer.scala`.

Comment (saraiva132) [reply core/shared/src/main/scala/zio/ZManaged.scala]: Yeah I do not mind removing the changes in `ZLayer.scala`, was trying to extract common behaviour but although I understood ZLayer that might been unnecessary. I still have the impression provideLayer for `ZIO` and `ZManaged` should look quite similar. Are you positive `provideSomeManaged(layer.value.map(ev))` suffices ?

Comment (saraiva132) [reply core/shared/src/main/scala/zio/ZManaged.scala]: Looking at it @adamgfraser, your proposal looks quite clean. I was just wondering if this translation from one level to another could be moved to `ZLayer` but TBH I am fine with leaving that aside.

Comment (saraiva132) [reply core/shared/src/main/scala/zio/ZManaged.scala]: Also, I am unsure whether your proposal would work. Should it not look like something like this, for us to have proper R translation? 
``` 
for {
      r0 <- ZManaged.environment[R0]
      r1 <- layer.value.provide(r0)
} yield ev(r1)).flatMap(self.provide(_)
```

Comment (adamgfraser) [reply core/shared/src/main/scala/zio/ZManaged.scala]: That isn't necessary because calling `value` on a `ZLayer[R0, E, R]` returns a `ZManaged[R0, E, R]` so it already expresses the ability to partially provide the environment. I checked that this compiles. We could add a test for it if you want.

**Comment by saraiva132** [comment]
@adamgfraser I was not strong enough to write what seemed like useless tests xD
